Geometry overhaul part 3.03 (final part for 3.6)

Elv13 opened the pull request:
* Add the last planned placement function [DONE]
* Turn tooltip into a class [DONE]
* Add shape support to tooltips [DONE]
* Support "next to drawable" tooltip mode like Max OS X [DONE]
* Add more tooltip properties and theme options [DONE]
* <s>Remove awful.menu geometry code and use awful.placement</s> later
* <s>Add hot corner support</s> later
* <s>Find a way to get the "smart wibox" (wibox that resize itself to fit a widget and update its own shape) from radical into awful or wibox (not sure yet)</s> later
* Test all this [DONE]

What wont make it (from the original TODO list of this PR serie):

 * Notification shape/placement (this require the dynamic layout system and I wont push it into 3.6)


Closing note:

This is the last PR of the geometry overhaul. After this PR, my Radical module will be 3K line shorted than the 3.5 version. My collision module will be much smaller too (partly because of the shape API too). I think so far it went well. Beside the placement code and the doc, awful also got smaller, but not by much. The first 4 PR were quite big and intrusive, but there was only a few regression reports so far. `awful.placement` may be a huge pile of complex and coupled/spaghetti code, but (I think/hope) it made everything else simpler and less coupled. Plus, this serie gave many cool new features I hope users will love. I hope this code will be the basis of tons of Awesomeness.

blueyed commented on lib/awful/mouse/init.lua:
```diff
@@ -326,6 +325,19 @@ end
 
+--- Get the current widget geometry.
+-- @property current_widget_geometry
+-- @tparam ?table The geometry.
```
s/tparam/treturn ?!

Elv13 replied:
properties documentation is using param. ldoc is configured to turn that into a "type" section. I could create an alias to make that less confusing...

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +617,150 @@ local function rect_to_point(rect, corner_i, corner_j)
+-- Create a pair of rectangles used to set the relative areas.
+-- v=vertical, h=horizontal
+local function get_cross_sections(abs_geo, mode)
+print("\n\nLA!",mode)
```
Leftover print.

psychon replied:
Still a leftover print a week later and when "WIP" is removed. I'll stop reading commits here and wait for you to specifically say "I handled all existing comments, please review again".

psychon replied:
Uhm. Github GUI says that this is an outdated diff. However, I reviewed this by `git pull`ing your branch and looking at this locally and noticed the `print` there independently from Daniel. What's going on?!?

Elv13 replied:
The fixup was accidently rebased in the wrong commit. I reviewed the pr as
a whole, so i missed it. Github ui fail quite hard to make that obvious.
On Jun 3, 2016 3:37 AM, "Uli Schlachter" <notifications@github.com> wrote:

In lib/awful/placement.lua
<https://github.com/awesomeWM/awesome/pull/929#discussion_r65667165>:

> @@ -603,6 +617,150 @@ local function rect_to_point(rect, corner_i, corner_j)
>      }
>  end
>
> +-- Create a pair of rectangles used to set the relative areas.
> +-- v=vertical, h=horizontal
> +local function get_cross_sections(abs_geo, mode)
> +print("\n\nLA!",mode)

Uhm. Github GUI says that this is an outdated diff. However, I reviewed
this by git pulling your branch and looking at this locally and noticed the
print there independently from Daniel. What's going on?!?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/awesomeWM/awesome/pull/929/files/405f522b13aa51399e6088c39ecd71086619684f#r65667165>,
or mute the thread
<https://github.com/notifications/unsubscribe/AAUxoNo82wOBni-FNogX9IBBW3F2uQ-Xks5qH9mdgaJpZM4IpNLU>
.

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +617,150 @@ local function rect_to_point(rect, corner_i, corner_j)
+        return {h=abs_geo, v=abs_geo}
+    end
+
+    assert(false)
```
Include the invalid mode here?

Elv13 replied:
This is a local function, I will remove it once the the WIP todos are done. This is still work in progress Not all features from the prototype (and Radical.tooltip) have been imported yet.

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +617,150 @@ local function rect_to_point(rect, corner_i, corner_j)
+    elseif (not geo.drawable) and geo.x and geo.width then
+        local coords = capi.mouse.coords()
+
+        -- Check id the mouse is in the rect
```
s/id/if

Elv13 commented:
I will stop this here. There is already too much features and breakages from me. Of course, you can't change the API as much as I did without breaking some eggs. I am sorry about that. I really believe the work I did to standardize and document the API will be beneficial for the project. If I had to do it again, I probably would not. It took _a lot_ of time, but addressed issues that have been bugging me for a long time. It also added some "cool" new features and flexibility. But this endeavor was way too large for a single release, let alone with 1200 other commits since 3.5.0.

So I am done for this release. This is it. I will fix whatever regressions I caused (that are not due to fragile hacks in modules/configs), but I wont submit new code until 3.7 is branched.

mindeunix commented:
@Elv13 Good work, as always.

Elv13 commented:
> @Elv13 Good work, as always.

Thanks. This PR doesn't fully merge Radical tooltips and I didn't merge the other Blind PRs after all. I will turn them back into modules. At least, we (/me + @mindeunix) will have a better base to build our configs. I once this is merged, I will push a Radical update for the recent API changes. This will drop some files, but it's not quite enough to drop smart_wibox, hot_corner and tooltip Radical components as I hoped.

psychon commented on lib/awful/mouse/init.lua:
```diff
@@ -326,6 +325,19 @@ end
+    local _, ret = mouse.object.get_current_widget()
+
+    return ret
+end
```
This handles `get_current_widget`. What about `get_current_widgets` (plural)?
Perhaps it is easier to make `get_current_widgets`  just return the result from `:find_widgets` instead of extracting the individual widgets (and perhaps renaming it to add `_geometries`)? I don't know.

psychon commented on lib/awful/mouse/init.lua:
```diff
@@ -326,6 +325,19 @@ end
+    return ret
+end
+
+function mouse.object.set_current_widget_geometry() end
```
Hm. Perhaps we should try getting rid of these "doing-nothing-setters" (and yes, this new idea doesn't belong into this PR).
What do you think about not allowing properties if there is a getter but not a setter in `gears.object.properties`?
Something like the following (untested):
```diff
diff --git a/lib/gears/object/properties.lua b/lib/gears/object/properties.lua
index 9b98868..a35239d 100644
--- a/lib/gears/object/properties.lua
+++ b/lib/gears/object/properties.lua
@@ -83,6 +83,12 @@ function object.capi_index_fallback(class, args)
             return args.setter_fallback(cobj, prop, value)
         end
 
+        -- If a getter exists but not a setter, then the property is read-only
+        if args.getterer_class and args.getter_class[getter_prefix..prop] then
+            -- error("Tried to set read-only-property (psychon always adds error()s, Elv13 doesn't like them")
+            return
+        end
+
         local fallback = properties[cobj] or cobj_register(cobj)
 
         -- Use the fallback property table
```

blueyed replied:
JFI: `args.getterer_class` seems wrong in the patch from @psychon ?!

blueyed commented on lib/awful/placement.lua:
```diff
@@ -552,8 +566,9 @@ attach = function(d, position_f, args)
-    -- If there is a parent drawable, screen or mouse, also track it
-    if parent then
+    -- If there is a parent drawable, screen, also track it
+    -- note that tacking the mouse is not supported
```
Please use punctuation.  These appear to be two comments, right?!

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+    assert(false)
+end
+
+-- When a rectangle is embedded into a bigger one, get the regious around
```
s/regious/region(s)/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+end
+
+-- When a rectangle is embedded into a bigger one, get the regious around
+-- the outline of the bigger rectangle closest to the smaller one (on each sides)
```
s/sides/side/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+
+    abs_widget_geo.drawable_geo = geo.drawable and dpos or geo
+
+    -- Get the comparaison point
```
"Get the point for comparison."

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+        bottom = {x = cs.v.x           , y = cs.v.y+cs.v.height},
+    }
+
+    -- Assume the section is part of a single screen until someone complain.
```
s/complain/complains/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+    }
+
+    -- Assume the section is part of a single screen until someone complain.
+    -- It is much faster to compute and getting it wrong probably have no side
```
s/have/has/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+    return regions
+end
+
+-- Check if the proposed geometry fit in the screen
```
s/fit in the screen/fits the screen/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -603,6 +618,160 @@ local function rect_to_point(rect, corner_i, corner_j)
+
+    local geo2 = region:get_rectangle(0)
+
+    -- If the geometry is the same, then it fit, else, it will be cropped
```
s/fit/fits/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -1143,9 +1322,96 @@ function placement.scale(d, args)
----@DOC_awful_placement_maximize_vertically_EXAMPLE@
+--- Move a drawable to a relative position next to another one.
+--
+-- The `args.preferred_positions` looks like:
```
s/looks/look/

blueyed commented on lib/awful/placement.lua:
```diff
@@ -1143,9 +1322,96 @@ function placement.scale(d, args)
+
+    local pref_idx, pref_name = 99, nil
+
+    local mode,wgeo = args.mode
```
Remove newlines between the assignments?!

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
 local tooltip = { mt = {}  }
 
-local instance_mt = {}
+-- The mouse point is 1x1, so anything aligned besed on it as parent
```
s/besed/based/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
-local instance_mt = {}
+-- The mouse point is 1x1, so anything aligned besed on it as parent
+-- geometry will go out of bound. To get the desired placement, it is
+-- necessary to swap left<-> and top<->bottom
```
s/swap left<-> and top<->bottom/swap left with right and top with bottom/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+    bottom_right = "top_left",
+    top          = "bottom",
+    bottom       = "top",
+}
```
This could use string substitution maybe?! (but feel free to ignore)

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+-- @beautiful beautiful.tooltip_opacity
+
+--- The tooltip shape.
+-- @beautiful beautiful.tooltip_shape
```
"shape" is not very specific.  Could need some more details.

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+    cr:fill()
+    wb.shape_bounding = img._native
+
+    -- The wibox background uses ARGB32 border so tooltip anti-aliasing work
```
s/work/works only/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+
+    -- The wibox background uses ARGB32 border so tooltip anti-aliasing work
+    -- when an external compositor is used. This will look better than
+    -- the capi.drawin own border support.
```
s/capi.drawin/capi.drawin's/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+    img = cairo.ImageSurface(cairo.Format.ARGB32, w, h)
+    cr  = cairo.Context(img)
+
+    -- Draw the border ( multiply by 2, then mask the inner part to save a path)
```
s/( /(/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+    cr  = cairo.Context(img)
+
+    -- Draw the border ( multiply by 2, then mask the inner part to save a path)
+    -- the -1
```
Leftover?

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,160 @@
+
+    local align  = self._private.align
+
+    local real_placement = align_convert[align]
```
Remove newlines between assignments.

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+    wb:set_widget(self.marginbox)
+
+    -- Close the tooltip when clicking it.  This gets done on release, to not
+    -- emit the release event on an underlying object, e.g. the titlebar icon.
```
Awesome comment! :cherries:

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+    return wb
+end
+
+--- If the tooltip is visible.
```
"Is the tooltip visible?"

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+end
+
+--- Set the tooltip positioning mode.
+-- This affect how the tooltip is placed. By default, the tooltip is `align`ed
```
s/affect/affects/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+
+--- Set the tooltip positioning mode.
+-- This affect how the tooltip is placed. By default, the tooltip is `align`ed
+-- close to the mouse cursor. It is also possible place the tooltip relative to
```
s/possible place/possible to place/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+
+--- The preferred positions when in `outside` mode.
+--
+-- If the tooltip fits on different side of the drawable, then pick the
```
s/different side/multiple sides/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+--- The preferred positions when in `outside` mode.
+--
+-- If the tooltip fits on different side of the drawable, then pick the
+-- priorities.
```
s/then pick the priorities/then this defines the priority/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -110,46 +244,190 @@ end
+--    {"top", "right", "left", "bottom"}
+--
+-- @property preferred_positions
+-- @tparam table preferred_positions The position, orderred by priorities
```
s/orderred/ordered/

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -170,11 +449,21 @@ end
 end
 
+-- Tooltip can be applied to both widgets, wibox and client, their geometry
+-- work differently.
```
s/work/works/

blueyed commented on tests/test-tooltip.lua:
```diff
@@ -0,0 +1,124 @@
+    ontop   = true
+}
+
+w: setup {
```
This looks weird with the space inbetween.

blueyed commented:
Left some comments and tried it actually.

It breaks both my custom tasklist_update, which uses the private `tb._layout` (which is arguably a fault on my side), but also the tooltips for my widgets do not show anymore.

I only see "SHOW" through https://github.com/awesomeWM/awesome/blob/42382bb5f6dfdb27cdc2b92637110909973bad47/lib/awful/tooltip.lua#L529-L535.

actionless commented:
> Notification shape/placement (this require the dynamic layout system and I wont push it into 3.6)

that's really sad, because i like how it looks (made with smart_wibox from radical):

![2016-06-04-113645_3840x2160_scrot](https://cloud.githubusercontent.com/assets/1655669/15798858/be2d6a04-2a48-11e6-9071-fe66fc9cd253.png)

also there is some minor issue with inconsistency between clipping and border shape but that's out of the scope of this conversation

Elv13 commented:
>  but the text is missing.

@blueyed Can you investigate a bit. All I can say is that the tests display text and the titlebar tooltips also display correctly. Given your config isn't public, there is little else I can do.

> that's really sad, because i like how it looks (made with smart_wibox from radical):

Sad, yes, but very time consuming to fix. Feel free to do it yourself. Your screenshot is a fine example of why I don't want to upstream those parts yet. You see the black pixels in the corners? Those would go away if everything was properly implemented (and a compositing manager was used). Getting anti-aliased border (both external and internal) wont be that easy. In fact, it is the same issue as #954. Wiboxes should use a `background` widget and drop the internal border support. However, the list of nasty side effects that would have to be fixed (including tons of geometry code) is too much for now.

blueyed commented:
@Elv13 
The tooltip also appears empty (or with the same fg/bg color) for the default titlebar with the icons there.
It is related to my custom theme however, which is derived from the xresources theme.  This is affected by the same issue.

Elv13 commented:
Can you set beautiful.tooltip_bg and fg to confirm this? Also, is it a regression? what was the color before?

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -244,27 +550,26 @@ tooltip.new = function(args)
         self.timer:connect_signal("timeout", self.timer_function)
     end
 
+    local fg = beautiful.tooltip_fg or beautiful.fg_focus or "#000000"
```
@Elv13 
This breaks for the xresources theme: it seems to use the same fg as bg.
Forcing this to be "#000000" makes the text of tooltips appear.

blueyed commented on lib/awful/tooltip.lua:
```diff
@@ -244,27 +550,26 @@ tooltip.new = function(args)
-        border_color = beautiful.tooltip_border_color or beautiful.border_normal or "#ffcb60",
+        border_width = 0,
+        fg = fg,
+        bg = color.transparent,
```
@Elv13 
I have also tried using "#000000" here, but it was not applied.

blueyed commented:
@Elv13 
> Can you set beautiful.tooltip_bg and fg to confirm this?

Yes, forcing this to red and green works.

> Also, is it a regression? what was the color before?

Based on your xrdb settings.

"fg_color" is still mentioned there: https://github.com/Elv13/awesome-1/blob/geometry_overhaul_p3.03/themes/xresources/theme.lua#L40 - but it does not seem to be used?! (assets: https://github.com/Elv13/awesome-1/blob/geometry_overhaul_p3.03/themes/xresources/assets.lua).

blueyed commented:
@actionless 
Does the xresources theme work for you with this PR?

actionless commented:
yup, tooltips are black on black then

i think in order to fix that we should just update the theme to have tooltip_{f,b}g explicitly defined

Elv13 commented:
@actionless or @blueyed Can you provide a patch or propose a change inline here. Which xresources color do you want for the tooltip fg and bg?

blueyed commented:
@actionless 
Can you look into it, please?
I am using some workaround, but am not sure what exactly.

blueyed commented:
@Elv13 
Shouldn't the defaults cover this use case also, or is the xresources theme weird/wrong in this regard?

Elv13 commented:
![cpu1](https://cloud.githubusercontent.com/assets/340384/16428548/627dfe7a-3d40-11e6-8c9a-614f3bc52ce9.png)

They work in all my themes and the default one. It "should" work with all themes with `fg_focus` set. However I just noticed that the old code use `bg_focus` as the fallback while the new one use `bg_normal`, so this part is wrong on my side (so is not correctly handle the old color name)

blueyed commented:
So, given that the xresources-theme gets adjusted I think this can be merged then, too?!

blueyed commented:
There are still some unaddressed comments in the file/diff view.
Assigning to @Elv13.

Elv13 commented:
Rebased, the comments have been addressed

psychon commented on lib/awful/tooltip.lua:
```diff
@@ -189,14 +483,28 @@ end
+
+    self:add_signal("property::visible")
+    self:add_signal("property::align")
+    self:add_signal("property::mode")
```
I haven't looked at this at all. I just saw a metric ton of warnings in the travis build (the LuaJIT-build failed because of `tests/test-screen-changes.lua`, our unstable test, running into a timeout).
Since commit 0857f6f1b5535a862c840311f5bdad849f42e95d, `add_signal` doesn't do anything except for printing a warning, so these three calls can be removed.

Elv13 commented:
I removed the `add_signal`s. Is this ok now? I am waiting for this before pushing some commits in my modules.

psychon commented on lib/awful/mouse/init.lua:
```diff
@@ -319,6 +318,32 @@ function mouse.object.get_current_widget()
     end
 end
 
+function mouse.object.set_current_widget() end
```
Empty setters are no longer necessary since cdf16d4660d73c7db988723aaee40076f7eaa9fe. Please remove (all of them).

psychon commented on lib/awful/tooltip.lua:
```diff
@@ -441,7 +527,10 @@ function tooltip.new(args)
         end
     else
-        function self.show()
+        function self.show(other, geo)
```
I don't really understand how this works. What type of object are `other` and `geo` here?

psychon commented:
I looked through this and left my comments in the form of fixup commits. I had to do a rebase so should I could add a fixup commit in the right place (commit A had a `print`, commit B removed it; I wanted to remove it from commit A). So, this is now rebased on lastest master and commit 3831d29cd69d34a7a2f0c93 is gone.

Please look through my changes, decide if they are good and then apply them if you are agree (`git checkout geometry_overhaul_p3.03 ; git rebase -i --autosqash master`). In case this whole fixup commits idea is wrong: `git reset --hard 0a9ea10710b928975bf4767d4b5a8f1ce94dcf25` (this is the branch before I changed something).

Please look at the commit message from 1c492ec96b86fb41068d6823b8b1eb23999f4b6e and 3555a26c2cb914d2e2ce6af72ab2d1a1edb5060d, because I am not 100% sure about these changes.

The only commit that I left here at github and that is still relevant: How are the arguments to `self.show` supposed to work? What is their type?

Elv13 commented:
Everything looks alright. Thanks!

> The only commit that I left here at github and that is still relevant: How are the arguments to self.show supposed to work? What is their type?

See `mouse::enter` arguments. It is a little magic, I will give you that.

Feel free to merge this.

blueyed commented:
Before merging this should be interactively rebased.. (to remove the fixup commits).

psychon commented:
> See mouse::enter arguments. It is a little magic, I will give you that.

On a drawin, there are no arguments.
On a wibox, there are no arguments.
On a widget, it's a table with entries `x, y, width, height, drawable, widget`.

None of this has two arguments and I cannot see how the code you add in this PR can work on anything but widgets.

Elv13 commented:
> None of this has two arguments and I cannot see how the code you add in this PR can work on anything but widgets.

But does it even make sense to use those on anything but widgets? Why would you want a tooltip on a wibox directly?

psychon commented:
I bet someone can come up with a reason why they want a tooltip on their conky window.
Anyway, changing this code to work only on widgets should be documented better. The code looks like it tries to work on multiple types of objects. E.g. the whole `get_parent_geometry` exists, but serves no purpose...? Also, the API docs talk only about "object", not "widget".

How about changing `add_to_object` so that things work on all kind of objects? It would require some magic to detect the kind of argument and creating some new function with the object as upvalue. (which makes `remove_from_object` harder to implement...)
For example:
```lua
function tooltip:add_from_object(obj)
    if self._private.connected_objects[obj] then return end
    self._private.connected_objects[obj] = {
        show = function(...) self.show(obj, ...) end,
        hide = function(...) self.hide(obj, ...) end,
    }
    obj:connect_signal("mouse::enter", self._private.connected_objects[obj].show)
    obj:connect_signal("mouse::leave", self._private.connected_objects[obj].hide)
end
function tooltip:remove_from_object(obj)
    if not self._private.connected_objects[obj] then return end
    obj:disconnect_signal("mouse::enter", self._private.connected_objects[obj].show)
    obj:disconnect_signal("mouse::leave", self._private.connected_objects[obj].hide)
    self._private.connected_objects[obj] = nil
end
```

Elv13 commented:
Rebased and fixed. I also have a new commit for anti-aliased tooltips to fix the issue in @actionless screenshot. I didn't add it to the PR and will probably submit it once it is merged. The current tooltip isn't affected by the issue, but lack outer anti-aliasing.

Elv13 commented:
@psychon Is the changes I made are good enough for this to be merged?

psychon commented on lib/awful/mouse/init.lua:
```diff
@@ -319,6 +318,28 @@ function mouse.object.get_current_widget()
+end
+
+--- Get the current widget geometries.
+-- @property current_widget_geometriesgeometriesy
```
I cannot really call this a typo, but only because it is worse than that. :-P

psychon commented on lib/awful/tooltip.lua:
```diff
@@ -83,7 +86,87 @@ local offset = {
+local function apply_shape(self)
+    local s = self._private.shape
+
+    if not s then return end
```
I think this has to do things like `self.wibox.shape_bounding = nil` and and `self.wibox:set_bg_image(nil)` (or something like this, I'm not completely sure).

Put differently: Does unsetting the shape work correctly?

psychon commented on lib/awful/tooltip.lua:
```diff
@@ -37,41 +37,162 @@
+        honor_workarea      = true,
+    })
+
+    if position ~= self._private.direction then
```
Correct me if I'm wrong, but `self._private.direction` is always `nil`?

psychon commented:
The above comments are just minor things (and I will soon be offline for another week), so feel free to merge this after you are done.
